### DISCUSSION TITLE
Persistent store deleting itself(?)

### DISCUSSION THREAD

**Discussion by A-PS1999** [pmndrs/zustand#1748] [Q&A]
Hey, I've been stumped by a problem I am having with one of my Zustand stores for a while now and I'd appreciate any insight any of you more experienced folks could provide in pointing out what I'm doing wrong and how to fix it.

I have a store set up like so using the persist middleware:

```
import { create } from 'zustand';
import { persist } from 'zustand/middleware';

export const useCartStore = create(
    persist((set) => ({
        cart: [],
        lastCartUpdate: null,
        updateItemCount: (index, count) => set((state) => {
            const cart = [...state.cart];
            cart[index].count += count;
            state.lastCartUpdate = Date.now();
            return cart;
        }),
        addItem: (id, count) => set((state) => {
            const alreadyInCart = state.cart.findIndex((item) => item.id === id);

            if (alreadyInCart === -1) {
                state.cart = [...state.cart, { id: id, count: count }],
                state.lastCartUpdate = Date.now();
            } else {
                state.updateItemCount(alreadyInCart, count);
            }
        }),
        removeItem: (id) => set((state) => ({ cart: state.cart.filter((item) => item.id !== id) })),
        clearCart: () => set({ cart: [], lastCartUpdate: null })
    }), { name: 'cart-storage' })
)
```

and I am using the store on one of my pages like so (simplified):

```
import React, { useState } from "react";
import { useCartStore } from "../../hooks/useCartStore.js";

const addItemSelector = (state) => state.addItem;

export default function ProductPage() {

    const addItem = useCartStore(addItemSelector);
    const [productDetails, setProductDetails] = useState([defaults to null, but can contain object of data]);
    const [quantity, setQuantity] = useState(1);

    const handleAddToCart = () => {
        addItem(productDetails.id, quantity);
    }

    return (
        <>
             <button type="button"  onClick={() => handleAddToCart()}></button>
        </>
    )
}
```

When I check Local Storage in my browser's Application tab I can see the state of the store like so:

`{state: {cart: [], lastCartUpdate: null}, version: 0}`

But when I press the button to call the `addItem` function my store seems to change and the cart is removed. This is what shows in the Local Storage tab:

`{state: {}, version: 0}`

This causes the page to crash because some of my other pages/components are trying to access the cart.
I have referred to the documentation and have tried some things like using Session Storage and reworking my code but my efforts haven't been successful. It's odd because this store used to work with an older version of Zustand (before you had to put braces around `create`) but now it's not working.

Sorry if this isn't the right place for this but I'm sure it's not an issue with zustand itself and I thought this'd be a good place to ask, given the number of people knowledgeable about zustand here.

**Comment by A-PS1999** [comment]
@dbritto-dev Thanks for your response on #1766. I don't want to hijack that discussion so I thought I'd continue my portion here.

Unfortunately, despite adding `createJSONStorage`, the store still gives me the same errors. See the below, for example:

```
import { create } from 'zustand';
import { persist, createJSONStorage } from 'zustand/middleware';

export const useCartStore = create(
    persist((set, get) => ({
        cart: [],
        lastCartUpdate: null,
        updateItemCount: (index, count) => set((state) => {
            const cart = [...state.cart];
            cart[index].count += count;
            state.lastCartUpdate = Date.now();
            return cart;
        }),
        addItem: (id, count) => set((state) => {
            const alreadyInCart = state.cart.findIndex((item) => item.id === id);

            if (alreadyInCart === -1) {
                state.cart = [...state.cart, { id: id, count: count }],
                state.lastCartUpdate = Date.now();
            } else {
                state.updateItemCount(alreadyInCart, count);
            }
        }),
        removeItem: (id) => set((state) => ({ cart: state.cart.filter((item) => item.id !== id) })),
        clearCart: () => set({ cart: [], lastCartUpdate: null })
    }), { name: 'cart-storage', storage: createJSONStorage(() => localStorage) })
)
```
Whether or not I use localStorage or sessionStorage, the result is the same. Similarly, I get the same error even if I replace every instance of `state` in the `addItem` action with `get()`.

**Reply by dbritto-dev** [reply]
@A-PS1999 let me look into that I'll create a demo for you
